Source/Dafny/Dafny.atg

```diff
@@ -2730,7 +2730,7 @@ VarDeclStatement<.out Statement/*!*/ s.>
        } else {
          update = new UpdateStmt(assignTok, endTok, lhsExprs, rhss);
        }
-       s = new VarDeclStmt(x, endTok, lhss, update);
+       s = new VarDeclStmt(x, endTok, lhss, update, isGhost);
```

Comment:
This seems suspicious to me. `isGhost` says whether or not the `ghost` keyword was used as part of the declaration. But this does not determine anything about ghostness of any initializing RHSs. For example, if the RHS is a method call (where all of the method's out-parameters are ghost), then

``` dafny
ghost var x, y, z := M(...);
```

declares 3 ghost variables and then performs a non-ghost method call. As an example in the other direction, suppose `N` is a non-ghost method with 3 out parameters, the first and third of which are ghost. Then, in

``` dafny
var x, y, z := N(...);
```

the local variables `x` and `z` will be "auto-ghosted". Hence, this statement declares 2 ghost variables and 1 non-ghost variable and then makes a non-ghost method call. As a third example, consider

``` dafny
var x, y, z := E, F, G;
```

where `E` and `G` and ghost expressions and `F` is a non-ghost expression. Here, `x` and `z` will again be auto-ghosted, so the statement declares 2 ghost variables, 1 non-ghost variable, and then performs a simultaneous assignment.

Reply:
You're right, and in the newest code, I actually had removed the code that was checking this "declared ghost" attribute., so I removed it.

Something interesting now is that, if we have the statement:
`var h := forall x: GhostSubsetType:: x in OriginalType ==> PredicateFailingOnNonGhostSubsetType(x);`
then h will be inferred to be ghost and thus it will accept the `GhostSubsetType` on the right.

However, it enable me to further think about this ghost inference and here is something that did not work before and that now works thanks to my last commit:
```dafny
method firstParameterGhost(ghost b: bool) {
}
function method firstParameterGhostFun(ghost b: bool) {
}
method Main() {
  firstParameterGhost(forall x: GhostSubsetType:: x in OriginalType ==> PredicateFailingOnNonGhostSubsetType(x));
  var x := firstParameterGhostFun(forall x: GhostSubsetType:: x in OriginalType ==> PredicateFailingOnNonGhostSubsetType(x));
}
```
Now it correctly infers the actual binding to be ghost and accepts the usual typing.